man/hp6038.1.in
.TH hp6038 1
.SH NAME
hp6038 \- control HP 603xA power supplies
.SH SYNOPSIS
.nf
.B hp6038 [\fIOPTIONS\fR]
.fi
.SH DESCRIPTION
\fBhp6038\fR controls the HP/Agilent 603xA power supply via its 
GPIB interface.
\fBhp6038\fR performs actions in the order they appear on the command line.
.SH OPTIONS
\fBhp6038\fR accepts the following command line options:
.TP
\fB\-n\fR, \fB\-\-name\fR name
Set the name used to look up the instrument address in
the config file.  The default is \fIhp6038\fR.
.TP
\fB\-a\fR, \fB\-\-address\fR address
Set the instrument address.
See gpib-utils.conf(5) for a description of address formats.
.TP
\fB\-c\fR, \fB\-\-clear\fR
Clear the instrument to its power-on state, disables its output,
and clears any foldback condition.
.TP
\fB\-l\fR, \fB\-\-local\fR
Enable the front panel.
.TP
\fB\-i\fR, \fB\-\-get-idn\fR 
Print instrument model.
.TP
\fB\-S\fR, \fB\-\-selftest\fR 
Run instrument selftest.
.TP
\fB\-I\fR, \fB\-\-iset\fR float
Set current in amps.
.TP
\fB\-V\fR, \fB\-\-vset\fR float
Set voltage in volts.
.TP
\fB\-o\fR, \fB\-\-out\fR 0|1
Enable/disable output.
.TP
\fB\-F\fR, \fB\-\-foldback\fR off|cv|cc
Enables or disables foldback protection which will disable power supply 
output if power supply switches to whatever mode, CV or CC, is defined
as the fold (error) condition [Default off].
Use the \fB\-\-clear\fR option to clear a foldback condition.
.TP
\fB\-s\fR, \fB\-\-samples\fR number
Select number of current and voltage samples to take [Default 0].
.TP
\fB\-p\fR, \fB\-\-period\fR period
Set sample period [Default 0s].
.SH FILES
@X_SYSCONFDIR@/gpib-utils.conf
.br
~/.gpib-utils.conf
.SH "SEE ALSO"
gpib-utils.conf(5)
.PP
"Operating Manual, Agilent 603xA Family Autoranging System DC 
Power Supplies", Sept. 2004.
.PP
"Service Manual, Autoranging System DC Power Supply, Agilent Models 
6033A and 6038A", Sept. 2000.


NEWS
=========================================================================
gpib-utils-1.5 Release Notes                                   2014-08-21
=========================================================================

Moving code from sourceforge to github:

* Stop tracking autotools build products, etc.

* Remove ChangeLog (use commit log and NEWS), add README.md

These changes were staged in 2011:

* Export libvxi11.so library for general use.

* libvxi11 API cleanup, portability fix to gpib.c [Alessio Bogani].

* New utilities
  - ibquery:  Generic query utility.

=========================================================================
gpib-utils-1.4 Release Notes                                   2009-02-02 
=========================================================================

* Added support for new devices:
  - ics8064:  ICS Electronics 8064 Ethernet to Relay interface
  - a34401:   Agilent 34401A/34410A/34411A DMM.
  - hp6038:   Hewlett-Packard 603xA System DC Power Supply
  - hp6626:   Hewlett-Packard 6625A/6626A/6628A/6629A System DC Power Supply
  - hp6632:   Hewlett-Packard 6632A/6633A/6634A System DC Power Supply

* Handle non-zero GPIB board index and secondary GPIB addresses.

* Changed ics8065 utility to more general purpose icsconfig utility.

* Added automake/autoconf infrastructure so gpib-utils compiles with and
  without linux-gpib.

* Refactored ICS, VXI-11, and LSD code into (non-installed) libraries.

=========================================================================
gpib-utils-1.3 Release Notes                                   2007-11-23 
=========================================================================

* VXI-11 support for devices behind a Ethernet-GPIB gateway such as
  ICS 8065, or with built-in Ethernet ports such as the Agilent 6000 
  series oscilloscopes.

* Added config file: /etc/gpib-utils.conf.  Place device addresses here
  and overridde with -a option on command line.  If linux-gpib is used,
  /etc/gpib.conf need only contain bus master device information.

* Added support for new devices:
  - a6032 Agilent 6000-series oscilloscopes
  - dg535 Stanford Research Systems DG535 digital delay/pulse generator 
  - sr620 Stanford Research Systems SR620 Universal Time Interval Counter
  - ics8065 ICS Electroncs 8065 Ethernet-GPIB Controller
  - hp3488 Hewlett-Packard 3488A Switch/Control Unit

* Removed support for old instruments that I no longer have access to
  and which I know of nobody using:
  - r5005 Racal Dana model 5005 DMM
  - hp3455 Hewlett-Packard 3455A DMM
  - tek2230 Tektronix 2230 oscilloscope (never really working)


src/hp6038.c
/* This file is part of gpib-utils.
   For details, see http://sourceforge.net/projects/gpib-utils.

   Copyright (C) 2008-2009 Jim Garlick <garlick.jim@gmail.com>

   gpib-utils is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.

   gpib-utils is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with gpib-utils; if not, write to the Free Software Foundation, 
   Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA */

#if HAVE_CONFIG_H
#include "config.h"
#endif
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <errno.h>
#include <string.h>
#include <unistd.h>
#include <libgen.h>
#if HAVE_GETOPT_LONG
#include <getopt.h>
#endif
#include <assert.h>
#include <sys/time.h>
#include <math.h>
#include <stdint.h>

#include "libutil/util.h"
#include "gpib.h"

#define INSTRUMENT "hp6038"

/* serial poll reg */
#define HP6038_SPOLL_FAU    1
#define HP6038_SPOLL_PON    2
#define HP6038_SPOLL_RDY    16
#define HP6038_SPOLL_ERR    32
#define HP6038_SPOLL_RQS    64

/* status reg. bits (STS? - also ASTS? UNMASK? FAULT?) */
#define HP6038_STAT_CV      1
#define HP6038_STAT_CC      2
#define HP6038_STAT_OR      4
#define HP6038_STAT_OV      8
#define HP6038_STAT_OT      16
#define HP6038_STAT_AC      32
#define HP6038_STAT_FOLD    64
#define HP6038_STAT_ERR     128
#define HP6038_STAT_RI      256

/* err reg values (ERR?) */
static strtab_t petab[] = {
    { 0, "Success" },
    { 1, "Unrecognized character" },
    { 2, "Improper number" },
    { 3, "Unrecognized string" },
    { 4, "Syntax error" },
    { 5, "Number out of range" },
    { 6, "Attempt to exceed soft limits" },
    { 7, "Improper soft limit" },
    { 8, "Data requested without a query being sent" },
    { 9, "Relay error" },
    { 0, NULL }
};

/* selftest failures */
static strtab_t sttab[] = {
    { 0,  "Passed" },
    { 4,  "External RAM test failed" },
    { 5,  "Internal RAM test failed" },
    { 6,  "External ROM test failed" },
    { 7,  "GPIB test failed" },
    { 8,  "GPIB address set to 31" },
    { 10, "Internal ROM test failed" },
    { 12, "ADC zero too high" },
    { 13, "Voltage DAC full scale low" },
    { 14, "Voltage DAC full scale high" },
    { 15, "Voltage DAC zero low" },
    { 16, "Voltage DAC zero high" },
    { 17, "Current DAC full scale low" },
    { 18, "Current DAC full scale high" },
    { 19, "Current DAC zero low" },
    { 20, "Current DAC zero high" },
    { 0, NULL }
};

static int _interpret_status(gd_t gd, unsigned char status, char *msg);
static int _readiv(gd_t gd, double *ip, double *vp);

char *prog = "";

static const char *options = OPTIONS_COMMON "lciSI:V:o:s:p:F:";

#if HAVE_GETOPT_LONG
#define GETOPT(ac,av,opt,lopt) getopt_long(ac,av,opt,lopt,NULL)
static struct option longopts[] = {
    OPTIONS_COMMON_LONG,
    {"address",         required_argument, 0, 'a'},
    {"verbose",         no_argument,       0, 'v'},
    {"local",           no_argument,       0, 'l'},
    {"clear",           no_argument,       0, 'c'},
    {"get-idn",         no_argument,       0, 'i'},
    {"selftest",        no_argument,       0, 'S'},
    {"iset",            required_argument, 0, 'I'},
    {"vset",            required_argument, 0, 'V'},
    {"out",             required_argument, 0, 'o'},
    {"foldback",        required_argument, 0, 'F'},
    {"samples",         required_argument, 0, 's'},
    {"period",          required_argument, 0, 'p'},
    {0, 0, 0, 0},
};
#else
#define GETOPT(ac,av,opt,lopt) getopt(ac,av,opt)
#endif

static opt_desc_t optdesc[] = {
    OPTIONS_COMMON_DESC,
    { "l", "local", "return instrument to front panel control" },
    { "c", "clear", "initialize instrument to default values" },
    { "i", "get-idn", "print instrument model" },
    { "S", "selftest", "run selftest" },
    { "I", "iset", "set current" },
    { "V", "vset", "set voltage" },
    { "o", "out", "enable/disable output (0|1)" },
    { "F", "foldback", "set foldback (off|cv|cc)" },
    { "s", "samples", "number of samples [0]" },
    { "p", "period", "sample period" },
    { 0, 0, 0 },
};

int
main(int argc, char *argv[])
{
    gd_t gd;
    int rc, c, print_usage = 0;
    int exit_val = 0;
    char tmpstr[64];
    int samples = 0;
    double period = 0.0;
    int showtime = 0;

    gd = gpib_init_args(argc, argv, options, longopts, INSTRUMENT,
                        _interpret_status, 0, &print_usage);
    if (print_usage)
        usage(optdesc);
    if (!gd)
        exit(1);

    while ((c = GETOPT(argc, argv, options, longopts)) != EOF) {
        switch (c) {
            OPTIONS_COMMON_SWITCH
                break;
            case 'l': /* --local */
                gpib_loc(gd);
                break;
            case 'c': /* --clear */
                gpib_clr(gd, 1000000); /* same as 'CLR' command */
                gpib_wrtf(gd, "OUT 0;RST\n");
                sleep(1);
                break;
            case 'i': /* --get-idn */
                gpib_wrtstr(gd, "ID?\n");
                gpib_rdstr(gd, tmpstr, sizeof(tmpstr));
                printf("%s\n", tmpstr);
                break;
            case 'S': /* --selftest */
                gpib_wrtstr(gd, "TEST?\n");
                gpib_rdstr(gd, tmpstr, sizeof(tmpstr));
                if (sscanf(tmpstr, "TEST %d", &rc) == 1)
                    printf("self-test: %s\n", findstr(sttab ,rc));
                else
                    printf("self-test: %s\n", tmpstr);
                break;
            case 'I': /* --iset */
                gpib_wrtf(gd, "ISET %s\n", optarg);
                break;
            case 'V': /* --vset */
                gpib_wrtf(gd, "VSET %s\n", optarg);
                break;
            case 'o': /* --out */
                gpib_wrtf(gd, "OUT %s\n", optarg);
                break;
            case 'F': /* --foldback */
                /* N.B. do all models accept OFF|CV|CC?  If not, cvt to 0|1|2 */
                gpib_wrtf(gd, "FOLD %s\n", optarg);
                break;
            case 's': /* --samples */
                samples = strtoul(optarg, NULL, 10);
                if (samples > 1)
                    showtime = 1;
                break;
            case 'p': /* --period */
                if (freqstr(optarg, &period) < 0) {
                    fprintf(stderr, "%s: error parsing period arg\n", prog);
                    fprintf(stderr, "%s: use units of: %s\n", 
                            prog, PERIOD_UNITS);
                    exit_val = 1;
                    break;
                }
                period = 1.0/period;
                break;
        }
    }

    if (samples > 0) {
        double t0, t1, t2;
        double i, v;

        t0 = 0;
        while (samples-- > 0) {
            t1 = gettime();
            if (t0 == 0)
                t0 = t1;
            if (_readiv(gd, &i, &v) == -1) {
                exit_val = 1;
                goto done;  
            }
            t2 = gettime();

            if (showtime)
                printf("%-3.3lf\t%lf\t%lf\n", (t1-t0), i, v);
            else
                printf("%lf\t%lf\n", i, v);
            if (samples > 0 && period > 0)
                sleep_sec(period - (t2-t1));
        }
    }

done:
    gpib_fini(gd);
    exit(exit_val);
}

static int
_interpret_status(gd_t gd, unsigned char status, char *msg)
{
    char tmpstr[64];    
    int e, err = 0;

    if (status & HP6038_SPOLL_FAU) {
        fprintf(stderr, "%s: device fault\n", prog);
        /* FIXME: check FAULT? and/or ASTS? but only if unmasked */
        err = 1;
    }
    if (status & HP6038_SPOLL_PON) {
        /*fprintf(stderr, "%s: power-on detected\n", prog);*/
    }
    if (status & HP6038_SPOLL_RDY) {
    }
    if (status & HP6038_SPOLL_ERR) {
        gpib_wrtstr(gd, "ERR?\n");
        gpib_rdstr(gd, tmpstr, sizeof(tmpstr));
        if (sscanf(tmpstr, "%d", &e) == 1)
            fprintf(stderr, "%s: prog error: %s\n", prog, findstr(petab, e));
        else
            fprintf(stderr, "%s: prog error: %s\n", prog, tmpstr);
        err = 1;
    }
    if (status & HP6038_SPOLL_RQS) {
        fprintf(stderr, "%s: device is requesting service\n", prog);
        err = 1; /* it shouldn't be unless we tell it to */
    }
    return err;
}

static int 
_readiv(gd_t gd, double *ip, double *vp)
{
    gpib_wrtstr(gd, "VOUT?\n");
    if (gpib_rdf(gd, "%lf", vp) != 1) {
        fprintf(stderr, "%s: error reading VOUT? result\n", prog);
        return -1;
    }
    gpib_wrtstr(gd, "IOUT?\n");
    if (gpib_rdf(gd, "%lf", ip) != 1) {
        fprintf(stderr, "%s: error reading IOUT? result\n", prog);
        return -1;
    }
    return 0;
}

/*
 * vi:tabstop=4 shiftwidth=4 expandtab
 */
